typed-racket-lib/typed-racket/private/parse-type.rkt

```diff
@@ -223,6 +237,36 @@
   (pattern (~and t:expr (~not :colon^) (~not :->^))
            #:when (not (syntax->list #'t))))
 
+;; TR type ASTs only allow identifiers either that
+;; a) are identifier-binding #t
+;; b) are generated by TR for internal use
+;;    (i.e. see genid and related functions in utils/utils.rkt)
+;; This function helps us create a fresh identifier
+;; that looks the same as the original and which meets the above
+;; conditions.
+(define (id->local-id id)
+  (gen-pretty-id (syntax->datum id)))
+
+(define-syntax-class dependent-fun-arg
+  #:description "dependent function argument"
+  #:attributes (name local-name deps type-stx refinement-prop-stx)
+  (pattern [name:id
+            :colon^
+            (~optional (dep:id ...)
+                       #:defaults ([(dep 1) null]))
+            type-stx:expr
+            (~optional (~seq #:where refinement-prop-stx:expr)
+                       #:defaults ([refinement-prop-stx #'Top]))]
+           #:fail-when (member #'name (syntax->list #'(dep ...)))
```

Comment:
I think this error-checking should be with the checks on 762. So the syntax class just does parsing, and all the checking happens in the same place.

Reply:
good point - fixed